### PR TITLE
Implement Google Cloud store using fsspec

### PR DESCRIPTION
**Pull request by simonbohnen**
## Checklist
* [x] Added a `docs/changes.rst` entry

## Issues with this PR
- [gcsfs](https://github.com/fsspec/gcsfs) doesn't escape file names correctly before inserting them into URLs
  - I'm currently doing this on our side, which results in different actual keys on GCS (can e.g. be verified with the gcloud CLI)
  - Submitted a [PR](https://github.com/fsspec/gcsfs/pull/502) to get this corrected in gcsfs according to the specifications by Google [here](https://cloud.google.com/storage/docs/objects#naming) and [here](https://cloud.google.com/storage/docs/request-endpoints#encoding)

### CODE REVIEW

**Comment by simonbohnen** [comment]
@janjagusch Could you take a look at this PR?

File: minimalkv/fsspecstore.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            return self.fs.open(f"{self.prefix}{escape(key)}")
+        except FileNotFoundError:
+            raise KeyError(key)
+
+    def _put_file(self, key: str, file: IO) -> str:
+        # fs.put_file only supports a path as a parameter, not a file
+        # Open file at key in writable mode
+        with self.fs.open(f"{self.prefix}{escape(key)}", "wb") as f:
+            copyfileobj(file, f)
+            return key
```
Comment (janjagusch) [inline minimalkv/fsspecstore.py:160]: I think you should be able to use [`pipe`](https://github.com/fsspec/filesystem_spec/blob/3969aaf7378198b92df8dee99cd464d272a0d0f4/fsspec/spec.py#L727-L747)

File: minimalkv/fsspecstore.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    path: str
+        The path to escape.
+
+    Returns
+    -------
+    str
+        The escaped path.
+    """
+    # Escape key before downloading
+    from urllib.parse import quote
```
Comment (janjagusch) [inline minimalkv/fsspecstore.py:184]: any good reason to put the import here?

File: minimalkv/net/gcstore.py [hunk 3]
```diff
@@ -55,16 +55,34 @@ def __init__(
...
     ):
+        if isinstance(credentials, str):
+            # Parse JSON from path to extract project name
+            try:
+                with open(credentials) as f:
+                    credentials_dict = json.load(f)
+                    project = project or credentials_dict["project_id"]
+            except (FileNotFoundError, json.JSONDecodeError) as error:
+                print(error)
+                pass
```
Comment (janjagusch) [inline minimalkv/net/gcstore.py:66]: Why do we print the error and then pass?

File: tests/test_gcloud_store.py [hunk 4]
```diff
@@ -104,9 +106,9 @@ def test_gcstore_pickling(store):
     assert store.get("key1") == b"value1"
 
 
-def test_gcstore_pickling_attrs():
+def test_gcstore_pickling_attrs(gc_credentials):
     store = GoogleCloudStore(
-        credentials="path_to_json",
+        credentials=gc_credentials,
```
Comment (janjagusch) [inline tests/test_gcloud_store.py:111]: Why was this change necessary?

File: environment.yml [hunk 5]
```diff
@@ -4,24 +4,31 @@ channels:
...
-  - boto3
-  - pymongo
-  - docker-compose
-  - pymysql
+# Infrastructure
   - pre-commit
   - setuptools-scm
-  - uritools
   - sphinx
+  - docker-compose
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply environment.yml:34]: Yeah I reordered them, just to make it clearer which dependencies are used for what. I can put them back though.

File: minimalkv/fsspecstore.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        Raises
+        ------
+        IOError
+            If there was an error accessing the store.
+        """
+        # List files
+        all_files_and_dirs = self.fs.find(f"{self.prefix}", prefix=escape(prefix))
+
+        # When no matches are found, the Azure FileSystem returns the container,
+        # which is not desired.
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:134]: It's not for the Google Cloud Store. Removed it for now 👍🏼

File: minimalkv/fsspecstore.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    def _delete(self, key: str):
+        try:
+            self.fs.rm_file(f"{self.prefix}{escape(key)}")
+        except FileNotFoundError:
+            pass
+
+    def _open(self, key: str) -> IO:
+        if not self._prefix_exists:
+            raise NotFound("Bucket does not exist.")
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:149]: I agree. I moved this to the gcstore, where we can be certain that the prefix specifies the bucket.

File: minimalkv/fsspecstore.py (hunk 1 above)
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:160]: Thx!

File: minimalkv/fsspecstore.py (hunk 2 above)
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:184]: Nope, moved it to the top 👍🏼

File: minimalkv/net/gcstore.py (hunk 3 above)
Comment (simonbohnen) [reply minimalkv/net/gcstore.py:66]: The pass is pointless 👍🏼

File: minimalkv/net/gcstore.py [hunk 8]
```diff
@@ -55,16 +55,34 @@ def __init__(
...
 
+        self.project_name = project
         self._credentials = credentials
         self.bucket_name = bucket_name
         self.create_if_missing = create_if_missing
         self.bucket_creation_location = bucket_creation_location
-        self.project_name = project
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/net/gcstore.py:63]: No 👍🏼

File: minimalkv/net/gcstore.py [hunk 9]
```diff
@@ -55,16 +55,34 @@ def __init__(
...
+    def _open(self, key: str) -> IO:
+        return cast(IO, FSSpecStoreEntry(super()._open(key)))
+
+    # These exist to allow the store to be pickled even though some properties don't support pickling.
+    # We make pickling work by omitting the lazy properties from __getstate__
+    # and just (re)creating them when they're used (again).
+    # `_bucket` and `_client` are used for testing only.
+    # All actual operations are performed using `_fs`.
     @lazy_property
     def _bucket(self):
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/net/gcstore.py:92]: I agree. I would suggest to continue using the Google Cloud Storage SDK directly for setting up the testing contexts, as that prevents testing `fsspec` against `fsspec`. I will move these methods to the test files.

File: tests/test_gcloud_store.py (hunk 4 above)
Comment (simonbohnen) [reply tests/test_gcloud_store.py]: The `fsspec` filesystem was created in the initializer, which requires correct credentials. I deferred it to the first access now, which is what the previous implementation did.

File: minimalkv/fsspecstore.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+            Number of bytes to be returned.
+
+        """
+        return self._file.read(size)
+
+    def seekable(self):  # noqa
+        return self._file.seekable()
+
+    def readable(self):  # noqa
+        return self._file.readable()
```
Comment (janjagusch) [inline minimalkv/fsspecstore.py:73]: any reason these are not type-hinted?

(the type hinting in this module generally seems to be a bit inconsistent)

File: minimalkv/fsspecstore.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    ----------
+    path: str
+        The path to unescape.
+
+    Returns
+    -------
+    str
+        The unescaped path.
+    """
+    return unquote(path)
```
Comment (janjagusch) [inline minimalkv/fsspecstore.py:209]: why wrap these simple function calls?

File: minimalkv/net/gcstore.py [hunk 12]
```diff
@@ -55,197 +20,36 @@ def __init__(
...
+        if isinstance(credentials, str):
+            # Parse JSON from path to extract project name
+            try:
+                with open(credentials) as f:
+                    credentials_dict = json.load(f)
+                    project = project or credentials_dict["project_id"]
+            except (FileNotFoundError, json.JSONDecodeError) as error:
+                print(
+                    f"Could not get the project name from the credentials file: {error}"
+                )
```
Comment (janjagusch) [inline minimalkv/net/gcstore.py:32]: is it okay for the `project` attribute to be `None`? Or will this mess with code execution further down the road?

File: minimalkv/net/gcstore.py [hunk 13]
```diff
@@ -55,197 +20,36 @@ def __init__(
...
-            if isinstance(file, io.BytesIO):
-                # not passing a size triggers a resumable upload to avoid trying to upload
-                # large files in a single request
-                # For BytesIO, getting the size is cheap, therefore we pass it
-                blob.upload_from_file(file_obj=file, size=file.getbuffer().nbytes)
-            else:
-                blob.upload_from_file(file_obj=file)
-        return key
+    def _create_filesystem(self):
+        from gcsfs import GCSFileSystem
```
Comment (janjagusch) [inline minimalkv/net/gcstore.py:43]: any reason that import is nested in the function body?

File: tests/test_gcloud_store.py [hunk 14]
```diff
@@ -64,6 +67,39 @@ def try_delete_bucket(bucket):
...
+    cached_client = getattr(gcstore, f"{LAZY_PROPERTY_ATTR_PREFIX}client", None)
+    if cached_client:
+        return cached_client
+
+    if type(gcstore._credentials) == str:
+        client = Client.from_service_account_json(gcstore._credentials)
+    else:
+        client = Client(credentials=gcstore._credentials, project=gcstore.project_name)
+
+    setattr(gcstore, f"{LAZY_PROPERTY_ATTR_PREFIX}client", client)
```
Comment (janjagusch) [inline tests/test_gcloud_store.py:99]: why are you setting the client (and the bucket, for that matter) as attributes inside the store?

**Review by janjagusch — COMMENTED** [review]
nice!

File: minimalkv/fsspecstore.py (hunk 10 above)
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:73]: True yeah. Made it more consistent.

File: minimalkv/net/gcstore.py (hunk 13 above)
Comment (simonbohnen) [reply minimalkv/net/gcstore.py:43]: I think I saw it somewhere else, and I kinda like it to indicate that the import is only used here. I'm fine with moving it to the top though.

File: tests/test_gcloud_store.py (hunk 14 above)
Comment (simonbohnen) [reply tests/test_gcloud_store.py:99]: I did that to avoid recreating the `Bucket` object every time. It's probably a relatively cheap operation, so maybe we can also skip that caching.

File: minimalkv/fsspecstore.py (hunk 11 above)
Comment (janjagusch) [reply minimalkv/fsspecstore.py:209]: i see, that makes sense. still, maybe a bit too much boilerplate for something so simple.

how do you like:

```py
from functools import partial
from urllib.parse import quote as _quote
from urllib.parse import unquote

quote = partial(_quote, safe="")
```

and then using `quote` and `unquote` instead of `escape` and `unescape`?

File: minimalkv/fsspecstore.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+            return self._file.seek(loc, whence)
+        except ValueError:
+            # Map ValueError to IOError
+            raise IOError
+
+    def tell(self):
+        """Return the current offset as int. Always >= 0."""
+        if self.closed():
+            raise ValueError("I/O operation on closed file.")
+        return self._file.tell()
```
Comment (janjagusch) [inline minimalkv/fsspecstore.py:58]: There is still a return type hint missing.

**Review by janjagusch — APPROVED** [review]
Amazing! Great improvement to the previous implementation concerning readability. I only have a couple of smaller unresolved comments.

File: minimalkv/fsspecstore.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+    def _create_filesystem(self) -> AbstractFileSystem:
+        # To be implemented by inheriting classes.
+        raise NotImplementedError
+
+    @lazy_property
+    def _fs(self) -> AbstractFileSystem:
+        fs = self._create_filesystem()
+
+        if self.mkdir_prefix:
+            fs.mkdir(self.prefix)
```
[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:168]: No, see [here](https://github.com/fsspec/gcsfs/blob/c186d378d7b0995864aa948505c2a2e89d244e71/gcsfs/core.py#L655)

[the remark this answers is not in the document]
Comment (simonbohnen) [reply minimalkv/fsspecstore.py:168]: Still safest to only create it if it doesn't exist 👍🏼

File: minimalkv/net/gcstore.py (hunk 12 above)
Comment (simonbohnen) [reply minimalkv/net/gcstore.py]: Yes, having the project name is important when running against live GCS, as it is required to create new buckets.
I added a warning message, because the bucket might already exist, which wouldn't be a problem.
